### DISCUSSION TITLE
Nested field array being removed from handle submit

### DISCUSSION THREAD

**Discussion by mmahalwy** [react-hook-form/react-hook-form#6025] [Q&A] [answered]
**Describe the bug**
When using nested field array, a change to the parent does not reflect down to the field array (seems to be caching with `useRef`). 

**To Reproduce**
- Go to: https://codesandbox.io/s/react-hook-form-reset-usefieldarray-forked-l35qp?file=/src/index.js
- Click delete on the first item. Notice the second item is now the first. 
- Click submit. Notice the console log has the `names` array with 1 item, with empty `children` array. Notice that the children now disappeared from the UI. 

**Codesandbox link (Required)**
https://codesandbox.io/s/react-hook-form-reset-usefieldarray-forked-l35qp?file=/src/index.js

V7

- [JS](https://codesandbox.io/s/great-paper-deove) Template
- [TS](https://codesandbox.io/s/eager-sun-jt4df) Template

**Expected behavior**
The nested `children` array should maintain it's values

**Screenshots**
N/A

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

**Comment by bluebill1049** [comment, marked as answer]
can you refer to the [examples](https://github.com/react-hook-form/react-hook-form/tree/master/examples)? see what do you missing? https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-m8w6j?file=/src/nestedFieldArray.js

**Reply by mmahalwy** [reply]
@bluebill1049 I dont exactly follow. What's missing?

**Reply by mmahalwy** [reply]
@bluebill1049 I was able to reproduce the bug in your example. See Loom video: https://www.loom.com/share/0a96e6a4c6294033a8c2e8ab6310cbc7

Once `reset` is used, it removed the nested field array. See this forked example with `reset` usage: https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-forked-nbid9?file=/src/index.js


Edit: seems like this happens only when reset is used within `onSubmit`

**Reply by mmahalwy** [reply]
Update 2: this seems like the only way to fix this https://codesandbox.io/s/react-hook-form-handlesubmit-with-reset-xu1zu?file=/src/index.js:0-42. For some reason, if `reset` is called in a callback (`onSubmit` or Apollo's `onComplete`), it doesn't work correctly. But in a `useEffect` it's fine.

**Reply by bluebill1049** [reply]
you should use reset in the useEffect.
https://react-hook-form.com/api/useform/reset
<img width="1002" alt="Screen Shot 2021-07-21 at 6 54 35 pm" src="https://user-images.githubusercontent.com/10513364/126460939-cc560b05-f1ac-4281-9bd8-065b4b138e48.png">

**Reply by mmahalwy** [reply]
@bluebill1049 curious, how come it needs to be in `useEffect`?

**Reply by bluebill1049** [reply]
I have updated the doc on this part. 

- Because onSubmit is async include its validation and reset inside the callback will intercept the result of formState update.
- We can't set value for native input during onReset event.

It's recommended and safer to do a reset after the `callback` from onReset or onSubmit is completed.
